sdk-core-protos/src/lib.rs

```diff
@@ -1325,42 +1331,142 @@ pub mod coresdk {
         DeserializeErr(#[from] anyhow::Error),
     }
 
-    // TODO: Once the prototype SDK is un-prototyped this serialization will need to be compat with
-    //   other SDKs (given they might execute an activity).
-    pub trait AsJsonPayloadExt {
-        fn as_json_payload(&self) -> anyhow::Result<Payload>;
+    /// JSON encoding for payload types
+    pub struct Json(());
+
+    impl<T: DeserializeOwned + Sized> Decoder<T> for Json {
+        type Error = serde_json::Error;
+
+        fn decode(payload: &Payload) -> Result<T, Self::Error> {
+            serde_json::from_slice(&payload.data)
+        }
     }
-    impl<T> AsJsonPayloadExt for T
-    where
-        T: Serialize,
-    {
-        fn as_json_payload(&self) -> anyhow::Result<Payload> {
-            let as_json = serde_json::to_string(self)?;
-            let mut metadata = HashMap::new();
-            metadata.insert(
-                ENCODING_PAYLOAD_KEY.to_string(),
-                JSON_ENCODING_VAL.as_bytes().to_vec(),
-            );
-            Ok(Payload {
-                metadata,
-                data: as_json.into_bytes(),
-            })
+
+    impl<T: Serialize + ?Sized> Encoder<T> for Json {
+        type Error = serde_json::Error;
+
+        fn encode(value: &T) -> Result<Payload, Self::Error> {
+            let mut metadata = HashMap::default();
+            metadata.insert(ENCODING_PAYLOAD_KEY.to_owned(), JSON_ENCODING_VAL.to_vec());
+            serde_json::to_vec(value).map(|data| Payload { metadata, data })
         }
     }
 
-    pub trait FromJsonPayloadExt: Sized {
-        fn from_json_payload(payload: &Payload) -> Result<Self, PayloadDeserializeErr>;
+    pub trait Encoder<T: ?Sized> {
+        /// Error type for encoding
+        type Error: std::error::Error + Send + Sync + 'static;
+
+        /// Encode a value of the given type into a [`Payload`]
+        fn encode(value: &T) -> Result<Payload, Self::Error>;
     }
-    impl<T> FromJsonPayloadExt for T
-    where
-        T: for<'de> Deserialize<'de>,
-    {
-        fn from_json_payload(payload: &Payload) -> Result<Self, PayloadDeserializeErr> {
-            if !payload.is_json_payload() {
-                return Err(PayloadDeserializeErr::DeserializerDoesNotHandle);
-            }
-            let payload_str = std::str::from_utf8(&payload.data).map_err(anyhow::Error::from)?;
-            Ok(serde_json::from_str(payload_str).map_err(anyhow::Error::from)?)
+
+    pub trait Decoder<T> {
+        /// Error type for decoding
+        type Error: std::error::Error + Send + Sync + 'static;
+
+        /// Decode a [`Payload`] into a value of the desired type
+        fn decode(payload: &Payload) -> Result<T, Self::Error>;
+    }
+
+    /// Trait for types that are used as workflow or activity output
+    pub trait ToPayload {
+        /// The encoding with which the payload is serialized
+        type Encoder: Encoder<Self>;
```

Comment:
> Reason being it locks types to a particular encoder forever and the user can't change it. EX: Later you define `PayloadExt` for `String` as using the `Json` encoder - but what if someone wants to encode strings using a binary format?

Downstream crates will be unable to implement traits this crate provides for types from std anyway, so I'm not sure how you see this working.

Also, I feel like the ideal developer experience would be that workflows/activities do not need to manually encode/decode, which kind of requires (as far as I can tell) that the types bind themselves to an encoder/decoder. So this is trying to maintain the (desirable, IMO) status quo that the choice of encoding is bound to the type (currently, JSON via serde), rather than relying on an explicit conversion to `Payload`.

The latter is also a valid design choice but IMO less attractive. For user-defined types, it seems unlikely that callers want to encode with more than one different schemes (and if they still do, handling that via a newtype seems straightforward enough).

> Then users provide a `DataConverter` to the worker. We should implement `To/FromPayload` automatically for anything `serde` serializable in conjunction with the user simply providing their serde backing (ex `serde_json`) for the actual format.

To me, specializing all of the conversions on serde also seems potentially problematic. What if a downstream user prefers to use rkyv (for its superfast direct to memory deserialization), or abomonation, or nanoserde? The current setup makes the serde case really pretty easy without imposing that everything is serde all the time.

Reply:
Yeah, they won't need to manually encode or decode -  the point is that they provide a `DataConverter` to the worker, and _all_ `Payloads` go through this. The type of encoding shouldn't be bound to the type being converted from/to, it should be bound to the `DataConverter`. Then anything that is serializable and being used as input / a return value gets run through the converter on the way out/in.

As for the problem with downstream crates not being able to implement `To/FromPayload` directly & not forcing always using serde... I can think of a few options.

One of them is to just handle all this at runtime, allow returning anything that can be turned into a `Box<dyn Any>`, then do casting and the checking to see if something is serializable at runtime. 

Second option is to have a special case for `serde`, and everything else just needs one extra line at the end of the workflow: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=76d0eafab777cfb157e00684a01c7493 -- I like this option a lot because the vast majority of people are going to be using / happy with serde, but it gives a way out for the people who want to do something special.

The third option, which I didn't get all the way compiling because it's pretty tricky and I didn't want to spend forever on this, but probably does work long run, is to parameterize workflow functions over the user-provided serializer/dataconverter. IE: You have something like:

```rust
fn my_workflow_fn<S: TemporalSerializer>() -> impl CanBeSerializedWith<S> {}
```

In the real future SDK, they wouldn't have to write that generic in the workflow function because we'd handle it with a macro - but they would get compile time breaks if the thing they return from a workflow can't be serialized with the serializer they provided when constructing the worker. I have a feeling this approach is overkill though.

Reply:
Will think about this some more, just asking for some clarification:



> Yeah, they won't need to manually encode or decode - the point is that they provide a `DataConverter` to the worker, and _all_ `Payloads` go through this. The type of encoding shouldn't be bound to the type being converted from/to, it should be bound to the `DataConverter`. Then anything that is serializable and being used as input / a return value gets run through the converter on the way out/in.

Do you mean that the `Worker` should some kind of `Box<dyn DataConverter>` which should then try to convert the input and output types to `Payload`? IMO using trait objects like this isn't really playing to Rust's strengths, because it would not allow compile-time checking that the `Worker`'s `DataConverter` can handle the type (unless you're okay with the worker taking a type parameter for the `DataConverter`?).

> In the real future SDK, they wouldn't have to write that generic in the workflow function because we'd handle it with a macro - but they would get compile time breaks if the thing they return from a workflow can't be serialized with the serializer they provided when constructing the worker. I have a feeling this approach is overkill though.

I'd like to soon start working on a workflow trait and an activity trait, because stringly typing workflow and actitivy types feels very error-prone. Can you sketch what you have in mind for how the macro would look?

Reply:
Yes, so that's why I like option 2 - you at least get compile time guarantees for `serde` which works for the vast majority of usecases. Option 3 does exactly what you said, parameterize the worker over the converter, and then in turn the workflows over the converter. I think that'd probably be a step too far. Compile time safe, but, annoyingly intrusive generics.


There is a PR here where I show (mostly) my ideas about the future workflow API https://github.com/temporalio/sdk-core/pull/550 - I definitely landed on wanting to use procmacros to help make things nice. I'm a little hesitant to encourage contribution here though because it's quite a large chunk of work, and we have a lot of lessons learned from other languages about what makes sense to do. Of course, I can see how annoying it is to say "don't do this we want to do it in house, but also we're not prioritizing it yet". So, if you're really into it, just be aware I might be slow to review and of course it is likely subject to change again later.